quill-core/src/main/scala/io/getquill/quotation/Parsing.scala

```diff
@@ -272,6 +273,15 @@ trait Parsing {
       case "!=" => EqualityOperator.`!=`
     }
 
+  val optionSingleOrderOperationParser: Parser[Operation] = Parser[Operation] {
+    case q"$a.nonEmpty" if is[Option[Any]](a) =>
+      BinaryOperation(astParser(a), EqualityOperator.`!=`, NullValue)
+    case q"$a.isDefined" if is[Option[Any]](a) =>
+      BinaryOperation(astParser(a), EqualityOperator.`!=`, NullValue)
+    case q"$a.isEmpty" if is[Option[Any]](a) =>
+      BinaryOperation(astParser(a), EqualityOperator.`==`, NullValue)
```

Comment:
We currently don't have cases of target languages that support optional values, but, for instance, we could support [queries on top of Scala collections](https://github.com/getquill/quill/issues/27), where `isDefined` has a completely different semantic from `!= null`. Also, I imagine that [DSH](https://github.com/getquill/quill/issues/156) has support for optional values.

Regarding `NullValue` being present in `quill-core`, it's a different case. It allows the DSL to produce an AST for quoted `null` values and the specific module is free to handle it or not. The important point is that having `NullValue` allows the quotation to express the quoted code without losing information, while making `isDefined` become `!= null` make it impossible to know if the quoted code was using `isDefined` or not.